Baikal/Kernels/CL/mipmap_level_scaler.cl

```diff
@@ -0,0 +1,261 @@
+/**********************************************************************
+Copyright (c) 2018 Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+********************************************************************/
+
+#ifndef MIPMAP_LEVEL_SCALER_CL
+#define MIPMAP_LEVEL_SCALER_CL
+
+#include <../Baikal/Kernels/CL/common.cl>
+#include <../Baikal/Kernels/CL/payload.cl>
+#include <../Baikal/Kernels/CL/utils.cl>
+#include <../Baikal/Kernels/CL/texture.cl>
+
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+
+KERNEL
+void ComputeWeights_NoRounding(
+    GLOBAL float3* restrict weights,
+    // Size of weight buffer
+    const int size)
+{
+    int id = get_global_id(0);
+
+    if (id < size)
+    {
+        weights[id].x = .0f;
+        weights[id].y = .5f;
+        weights[id].z = .5f;
+    }
+}
+
+KERNEL
+void ComputeWeights_RoundingUp(
+    GLOBAL float3* restrict weights,
+    // Size of weight buffer
+    const int size)
+{
+    int id = get_global_id(0);
+
+    float denominator = 2.f * size - 1.f;
+
+    // First weight
+    if (id == 0)
+    {
+        weights[id].x = .0f;
+        weights[id].y = ((float)size) / denominator;
```

Comment:
are you ready for NaNs?

Reply:
### *OH SHI—!*

Reply:
Actually, there shouldn't be any NaNs, as size should be greater than zero and denominator never equals zero.